code/game/objects/items/sentient_disease_injector.dm

```diff
@@ -0,0 +1,172 @@
+
+	update_icon()
+
+	return TRUE
+
+/obj/item/sentient_disease_injector/attack_self(mob/user)
+
+	if(!stored_vial)
+		return
+
+	stored_vial.forceMove(user.loc)
+	user.put_in_hands(stored_vial)
+	to_chat(user, span_notice("You eject \the [stored_vial] from \the [src]."))
+	stored_vial = null
+
+	update_icon()
+
+	return TRUE
+
+/obj/item/sentient_disease_injector/attack(obj/item/I, mob/user, params)
+	return //Prevents damage.
+
+/obj/item/sentient_disease_injector/afterattack(atom/target, mob/user, proximity)
+
+	if(!proximity || !iscarbon(target))
+		return ..()
+
+	if(uses <= 0)
+		to_chat(user, span_warning("\The [src] is spent!"))
+		return
+
+	var/mob/living/carbon/C = target
+
+	if(!C.can_inject(user, 1, user.zone_selected, TRUE))
+		to_chat(user, span_warning("You can't seem to inject \the [C] that way!"))
+
+	if(ishuman(C))
+		var/obj/item/bodypart/affecting = C.get_bodypart(check_zone(user.zone_selected))
+		if(!affecting)
+			to_chat(user, span_warning("\The [src] can't inject a missing limb!"))
+			return
+		if(affecting.status != BODYPART_ORGANIC)
+			to_chat(user, span_warning("\The [src] won't work on a robotic limb!"))
+			return
+
+	to_chat(user, span_notice("You stealthily inject \the [C] with \the [src]."))
+
+	uses -= 1
+
+	for(var/datum/disease/D as anything in C.diseases) //Cure all existing diseases
+		D.cure(add_resistance = FALSE)
+
+	if(stored_vial && stored_vial.reagents.total_volume) //If there is a stored vial, inject.
+		var/list/injected = list()
+		for(var/datum/reagent/R as anything in stored_vial.reagents.reagent_list)
+			injected += R.name
+		log_combat(user, C, "attempted to inject", src, "([english_list(injected)])")
+		stored_vial.reagents.reaction(C, INJECT, 1)
+		stored_vial.reagents.trans_to(C, stored_vial.reagents.total_volume, transfered_by = user)
+	else
+		log_combat(user, C, "attempted to inject", src)
+
+	if(length(C.diseases))
+		INVOKE_ASYNC(src, PROC_REF(create_sentient_virus), target, user)
+
+	update_icon()
+
+	return TRUE
+
+/obj/item/sentient_disease_injector/Destroy()
+	QDEL_NULL(stored_vial)
+	. = ..()
+
+/obj/item/sentient_disease_injector/proc/create_sentient_virus(mob/living/carbon/target,mob/user)
+
+	var/list/candidates = pollGhostCandidates("Do you wish to be considered for the special role of 'custom sentient disease'?", ROLE_ALIEN, null, ROLE_ALIEN)
+	if(!length(candidates))
+		return FALSE //No candidates.
+
+	var/mob/dead/observer/selected = pick_n_take(candidates)
+	var/mob/camera/disease/virus = new/mob/camera/disease(get_turf(target))
+	virus.key = selected.key
+	message_admins("[ADMIN_LOOKUPFLW(virus)] has been made into a sentient disease by [ADMIN_LOOKUPFLW(usr)]'s [src].")
+	log_game("[key_name(virus)] was spawned as a sentient disease by [ADMIN_LOOKUPFLW(usr)]'s [src].")
+	//Mix and cure all existing diseases.
+	for(var/datum/disease/D as anything in target.diseases)
+		if(D == virus)
+			continue
+		virus.disease_template.Mix(D)
+		D.cure(add_resistance = FALSE)
+
```

Comment:
Allow the sentient disease to pick a name via 	INVOKE_ASYNC(virus, TYPE_PROC_REF(/mob/camera/disease, pick_name))